### DISCUSSION TITLE
Filter 2d array?

### DISCUSSION THREAD

**Discussion by memark** [uiua-lang/uiua#204] [Learning] [answered]
The elements of a 1d array can easily be filtered with `▽ keep`
```
1_2_3_4
▽>2.
[3 4]
```

How can I do the same for a 2d array?

Things I've tried that don't work
```
[1_2_3 4_5_6]
▽>2.
Error: Keep amount must be a natural number or list of natural numbers, but its rank is 2
```
```
[1_2_3 4_5_6]
≡(▽>2.)
Error: Cannot append arrays of shapes [1 × 1] and [3]
```
```
[1_2_3 4_5_6]
>2.
≡▽
Error: Cannot couple arrays with shapes [1] and [3]
```

**Comment by kaikalii** [comment, marked as answer]
The problem becomes that the filtered rows may end up being different lengths, so you cannot reform the array normally.
The three main ways of solving this are `box`, `pack`, and `fill`.
- `box`ing the result of `keep` on each row wraps each filtered row in a box
- `pack` does the same, but only if it is necessary
- `fill` fills in the lengths of each row so that reforming the array works correctly

[Here](https://uiua.org/pad?src=QSDihpAgWzFfMl8zIDRfNV82XQpCIOKGkCBbMV8zXzQgMF80XzVdCiJCb3hlZCIKQm94ZWQg4oaQIOKJoSjilqHilr0pPjIuCkJveGVkIEEKQm94ZWQgQgoiUGFja2VkIgpQYWNrZWQg4oaQIOKKkOKJoeKWvT4yLgpQYWNrZWQgQQpQYWNrZWQgQgoiRmlsbGVkIgpGaWxsZWQg4oaQIOKsmjDiiaHilr0-Mi4KRmlsbGVkIEEKRmlsbGVkIEIK) are some examples of how they would be used here.
Notice the difference between using `box` and using `pack` on array `B`.

**Reply by memark** [reply]
Thank you so much! I keep forgetting about the need to box uneven-sized arrays. (In my mind I still only think about boxing when it comes to strings.) Super helpful!
